### ISSUE TITLE
Help using the models

### ISSUE THREAD

**Issue by ntelo007** [qubvel/segmentation_models#347]
Hi,

I would like to run some experiments and test the accuracy of all the available models on a semantic segmentation task. Does every model need to be build with a backbone architecture?

Furthermore, how can I run the pre-processing function you proposed without a backbone? 

Thank you for your time.

**Comment by JordanMakesMaps** [comment]
Every architecture (Unet, LinkNet, FPN, PSPNet) needs an encoder (resnet34, efficientnet-b7, densenet-201, etc.) so if you wanted to try out every permutation I would recommend a loop where for each architecture you load all of the encoders one at a time, have them train/test on your benchmark dataset, record the scores, clear the graph, and then let the loop continue for the rest of the encoders. Then do the same thing but for each of the architectures (each one requires different image dimensions so be sure to check those out).

So the `preprocess_input()` function is specific _to_ each backbone because the preprocessing in this repo is attempting to mimic what the original CNN was trained on for the ImageNet dataset. If you loaded up a ResNet50 with ImageNet weights, but then wanted to provide it with images where the range of pixel values were something like [-1, 1], it would wig out because it's only used to seeing images with pixels between [0, 255].

**Comment by ntelo007** [comment]
Doesn't the plain unet architecture have its own encoder? Without residual blocks ?

**Comment by JordanMakesMaps** [comment]
If you want to use a Vanilla U-Net with encoder check out [this ](https://github.com/karolzak/keras-unet) repo. All of the models is @qubvel's repo need to use one of the backbones as an encoder (alternatively you could also just create your own and splice it into the code).

**Comment by ntelo007** [comment]
I want to use this repository because I have already set my environment for that. Could you please recommend which combinations of networks and encoders are mostly used? Like unet with resnet 34 , etc.

**Comment by JordanMakesMaps** [comment]
@ntelo007 I've found that U-Net with the encoder frozen to be the best architectures. For encoders I've found that either of the DenseNets, and/or the EfficientNets (with 'noisy-student' weights) to be the best for my dataset.

I've played around with PSPNet, FPN and LinkNet, but none perform as well as the U-Net on my dataset.

**Comment by ntelo007** [comment]
What do you mean by frozen encoder? How can I match the right encoder with the LinkNet architecture or the PSPNet?

**Comment by JordanMakesMaps** [comment]
You can set the encoder of the model to be frozen when initiating the model. It means that none of the parameters within the encoder will be adjusted when training, just the parameters in the decoder. Alternatively, you could choose to properly fine-tune the model and unfreeze the encoder after a few epochs.

Trial and error is my best recommendation.

**Comment by twVolc** [comment]
> You can set the encoder of the model to be frozen when initiating the model. It means that none of the parameters within the encoder will be adjusted when training, just the parameters in the decoder. Alternatively, you could choose to properly fine-tune the model and unfreeze the encoder after a few epochs.

Is there a quick and easy way to unfreeze the encoder layers for fine tuning? I am freezing it when initiating the model, using `encoder_freeze=True`, but would then like to unfreeze it in later epochs as you suggest. I can't see a way to do this in the API - do I need to write my own loop through all of the layers setting them to `trainable=True`? It might be useful to have a method/function which can control this if it doesn't already exist.

**Comment by JordanMakesMaps** [comment]
I believe you're looking for:

```python
def set_trainable(model, recompile=True, **kwargs):
    """Set all layers of model trainable and recompile it
    Note:
        Model is recompiled using same optimizer, loss and metrics::
            model.compile(
                model.optimizer,
                loss=model.loss,
                metrics=model.metrics,
                loss_weights=model.loss_weights,
                sample_weight_mode=model.sample_weight_mode,
                weighted_metrics=model.weighted_metrics,
            )
    Args:
        model (``keras.models.Model``): instance of keras model
    """
    for layer in model.layers:
        layer.trainable = True

    if recompile:
        model.compile(
            model.optimizer,
            loss=model.loss,
            metrics=model.metrics,
            loss_weights=model.loss_weights,
            sample_weight_mode=model.sample_weight_mode,
            weighted_metrics=model.weighted_metrics,
        )
```

**Comment by twVolc** [comment]
Brilliant, thanks for such a quick reply. Not sure how I missed this!
